### ISSUE TITLE
One T4 is equivalent to eight logical cores of 6271 CPU about read_parquet permance test ?[QST]

### ISSUE THREAD

**Issue by chenrui17** [rapidsai/cudf#6620]
**What is your question?**
before in cudf-0.14 version , i tested cudf read_parquet api on T4 , and I also tested readTable of arrow lib api https://github.com/apache/arrow/blob/master/cpp/src/parquet/arrow/reader.h , and i found that cudf read_parquet on gpu is 8x better than cpu arrow read . it can be said that one T4 is equivalent to eight logical cores of 6271 CPU .
my question is that it's normal ? and below i will describe my test method and my test code , please help me and see if there is a mistake.
 
**test method**
to cudf ,I use read_parqeut c++ API , use threadPool to control cpu parallelism, and use rmm_pool to manage gpu memory , use PinedHostMemory to allocate host memory (cudaMallocHost not malloc)

**test data**
TPC-DS generate store_sales  table parquet file ,total 11 parquet files(each file is 1.2GB, and 11 columns int type, 12 columns float type) , if you need , i can transfer to you later 

**test results**
CPU: as shown is the red box (Tn indicates that the C++ thread pool has n threads enabled),  reading each 1.2GB parquet file the CPU arrow interface on average takes 7505ms in nvme mode , and 6110ms in  ramdisk mode.
![image](https://user-images.githubusercontent.com/33319780/97575421-a89f7000-1a27-11eb-847a-a7a90df43a47.png)
GPU: as shown is the red box, reading each 1.2GB parquet file the gpu cudf read_parquet API onaverage takes 1103ms in nvme mode, and 746ms in ramdisk mode.
![image](https://user-images.githubusercontent.com/33319780/97575430-adfcba80-1a27-11eb-991a-ed5bb2942406.png)

in short, under the condition of single thread, that is ,when the cpu uses at most one logical core, parquet file reading performance of a T4 card is 7-8X than that of cpu, which is approximately equivalent to 7-8 logical cores  of 6271 cpu (cause 7505/1103 = 6.8; 6110/746 = 8.2)

**code**
[read_parquet_cpu.txt](https://github.com/rapidsai/cudf/files/5458392/read_parquet_cpu.txt)
[read_parquet_gpu.txt](https://github.com/rapidsai/cudf/files/5458400/read_parquet_gpu.txt)

**Comment by chenrui17** [comment]
@jlowe  @karthikeyann @sameerz

**Comment by datametrician** [comment]
are the files compressed or uncompressed?

**Comment by chenrui17** [comment]
> are the files compressed or uncompressed?

default snappy compressed, I use this https://github.com/databricks/spark-sql-perf to generate parquet file

**Comment by datametrician** [comment]
We typically see better performance with larger files. For our TPCx-BB work we used 3GB partitions.
Also, I'd be interested to see this on V100 as well.
Has performance changed over versions, or do you just want it faster?

**Comment by chenrui17** [comment]
> We typically see better performance with larger files. For our TPCx-BB work we used 3GB partitions.
> Also, I'd be interested to see this on V100 as well.
> Has performance changed over versions, or do you just want it faster?

yeah , i expect it will has permance of one 6271 cpu. after all , other operators is very fast , like join and , groupby ,etc. can i think of read_parqeut performance is not good as other gpu operators ?
I also use python API to test read_parquet on V100 , and i found it behaves better than T4, compare to pandas API,  its performance is about 15 cores of 6271 cpu.

**Comment by datametrician** [comment]
I think read parquet is heavily dependent on the data, the partition size, and compression. For instance, with snappy going from 500mb to 2GB we see decompression go from 2GB/s to 6GB/s.

Please try a larger partition and let us know the perf, please.

**Comment by kkraus14** [comment]
How many row groups are in your parquet file and what page size is being used?

If I remember correctly, we want at least 1 row group per SM on the GPU and generally want small pages to expose more parallelism.

cc @vuule @devavret @nvdbaranec @OlivierNV who can advise better than myself 😄

**Comment by OlivierNV** [comment]
Generally, you want to least 8 pages per SM (default parquet page size is 1MB), more if there are wide differences in compression ratio between pages (often column type and data-dependent), in order to keep the GPU fully utilized.
There may be other things like the PCIE transfer rate of compressed data that can have a big impact (which is big chunk of read_parquet time) ~ One T4 being equivalent to 8 fast CPU cores for snappy decompression sounds in the right ballpark, though (lz decompression isn't particularly gpu-friendly).

**Comment by OlivierNV** [comment]
@chenrui17  It'd be interesting to also compare NvME numbers against the FILEPATH source type where io is done within cudf through mmap.

**Comment by kkraus14** [comment]
This was answered and has since gone stale. Closing.
